Implement

MythicManiac opened the pull request.

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+import sys
+import http.client
+import json
+from importlib._bootstrap import spec_from_loader
```
Underscored module import smells bad.

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+from importlib._bootstrap import spec_from_loader
+
+
+def get_suggestion(file, snippet, stops):
```
Type annotations?

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    headers = {
+        'OpenAI-Intent': 'copilot-ghost',
+        'OpenAI-Organization': 'github-copilot',
+        'Authorization': 'Bearer TOKEN',
```
It would be a good idea to make TOKEN a constant or parameter even if it's just TOKEN before you get a real token.

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        'Content-Type': 'application/json'
+    }
+    conn.request("POST", "/v1/engines/github-py-stochbpe-cushman-pii/completions", payload, headers)
+    res = conn.getresponse()
+    data = res.read()
```
No status code check?

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    return json.loads(data.decode("utf-8"))
+
+
+def get_fn(name):
```
annotations

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def get_fn(name):
+    args = get_suggestion(name + ".py", "def " + name, ["):\n"])["choices"][0]["text"]
+    full_snippet = "def " + name + args + "):\n"
```
prefer f-strings

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    args = get_suggestion(name + ".py", "def " + name, ["):\n"])["choices"][0]["text"]
+    full_snippet = "def " + name + args + "):\n"
+
+    code = get_suggestion(name + ".py", full_snippet, [
```
prefer f-strings

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        else:
+            return "\n".join(full_code.split("\n")[:block_ast.body[1].lineno - 1])
+
+    except:
```
don't do bare `except:`s ever

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    try:
+        block_ast = ast.parse(full_code)
+
+        if len(block_ast.body) == 1:
+            return full_code
+        else:
+            return "\n".join(full_code.split("\n")[:block_ast.body[1].lineno - 1])
```
what's the logic here?

MythicManiac replied:
in case copilot returns more than a single function, should ditch the extra code (or so I'm told)

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        raise ImportError("Copilot's plane crashed :(")
+
+
+def wrap_fn(name):
```
annotations

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    imports = []
+
+    def wrapper(*args):
+        try:
+            loc = {}
+            lines = source.split("\n")
+            exec(lines[0] + "\n    " + "\n".join(["import " + i for i in imports]) + "\n" + "\n".join(lines[1:]), globals(), loc)
+            return loc[name](*args)
+        except NameError as e:
+            missing_name = e.args[0].split("'")[1]
```
are you sure this is the minimal block for a NameError?

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+            missing_name = e.args[0].split("'")[1]
+
+            if missing_name in imports:
+                raise NameError("Copilot source references a missing value '" + missing_name + "'")
```
f-string

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+
+            if missing_name in imports:
+                raise NameError("Copilot source references a missing value '" + missing_name + "'")
+            else:
+                imports.append(missing_name)
+                return wrapper(*args)
```
what's the logic here? recursion smells bad

MythicManiac replied:
In case there's a name error, attempts re-run the code with the missing name added to imports first. In other words this should solve some cases where copilot forgets to add appropriate imports, though it could also be handled a lot better.

akx replied:
It should be noted in a code comment, not a GH comment

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+            else:
+                imports.append(missing_name)
+                return wrapper(*args)
+        except ModuleNotFoundError as e:
+            raise ImportError("Copilot source references a missing value that is also not a module")
```
e is unused, maybe `from e` in the raise?

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        return False
+
+
+sys.meta_path.append(CopilotImporter())
```
`def install()`

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+    return wrapper
+
+
+class CopilotImporter:
```
Is there a `typing` protocol or other ABC for meta_path members? If so, could use one

akx commented on copilot/__init__.py:
```diff
@@ -0,0 +1,116 @@
+        "n": 1,
+        "logprobs": 2,
+        "stop": stops,
+        "stream": False
```
```suggestion
        "stream": False,
```

akx reviewed (changes requested):
Missing a test/example that would explain how to use this (even if `TOKEN` is currently not available).

Maybe run things through black + isort for nice cleanliness?

MythicManiac committed:
Add pre-commit configuration

MythicManiac committed:
Add pyproject.toml

MythicManiac committed:
Add LICENSE

MythicManiac committed:
Add initial implementation

MythicManiac committed:
Update .gitignore

MythicManiac committed:
Rename package

MythicManiac committed:
Require explicit enabling of copilot import

Make it so the copilot.install method must be called before the magic
imports start working

MythicManiac committed:
Add README.md

MythicManiac committed:
Add comment explaining recursive wrap call
